apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,828 @@
+	// to this Gateway Listener. When unspecified or empty, the kinds of Routes
+	// selected are determined using the Listener protocol.
+	//
+	// A RouteGroupKind MUST correspond to kinds of Routes that are compatible
+	// with the application protocol specified in the Listener's Protocol field.
+	// If an implementation does not support or recognize this resource type, it
+	// MUST set the "ResolvedRefs" condition to false for this Listener with the
+	// "InvalidRoutesRef" reason.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Kinds []RouteGroupKind `json:"kinds,omitempty"`
+}
+
+// RouteSelectType specifies where Routes should be selected by a Gateway.
+//
+// +kubebuilder:validation:Enum=All;Selector;Same
+type RouteSelectType string
+
+const (
+	// Routes in all namespaces may be used by this Gateway.
+	RouteSelectAll RouteSelectType = "All"
+	// Only Routes in namespaces selected by the selector may be used by this Gateway.
+	RouteSelectSelector RouteSelectType = "Selector"
+	// Only Routes in the same namespace as the Gateway may be used by this Gateway.
+	RouteSelectSame RouteSelectType = "Same"
+)
+
+// RouteNamespaces indicate which namespaces Routes should be selected from.
+type RouteNamespaces struct {
+	// From indicates where Routes will be selected for this Gateway. Possible
+	// values are:
+	// * All: Routes in all namespaces may be used by this Gateway.
+	// * Selector: Routes in namespaces selected by the selector may be used by
+	//   this Gateway.
+	// * Same: Only Routes in the same namespace may be used by this Gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=Same
+	From *RouteSelectType `json:"from,omitempty"`
+
+	// Selector must be specified when From is set to "Selector". In that case,
+	// only Routes in Namespaces matching this Selector will be selected by this
+	// Gateway. This field is ignored for other values of "From".
+	//
+	// Support: Core
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// RouteGroupKind indicates the group and kind of a Route resource.
+type RouteGroupKind struct {
+	// Group is the group of the Route.
+	//
+	// +optional
+	// +kubebuilder:default=gateway.networking.k8s.io
+	Group *Group `json:"group,omitempty"`
+
+	// Kind is the kind of the Route.
+	Kind Kind `json:"kind"`
+}
+
+// GatewayAddress describes an address that can be bound to a Gateway.
+type GatewayAddress struct {
+	// Type of the address.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// +kubebuilder:default=IPAddress
+	Type *AddressType `json:"type,omitempty"`
+
+	// Value of the address. The validity of the values will depend
+	// on the type and support by the controller.
+	//
+	// Examples: `1.2.3.4`, `128::1`, `my-ip-address`.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Value string `json:"value"`
```

Comment:
Double check - do we really NEED type?

Reply:
One of our supported types is "NamedAddress" which we absolutely need a type for, I missed that initially. But since I was curious, here's what RFC 1123 says:

>          If a dotted-decimal number can be entered without such
>          identifying delimiters, then a full syntactic check must be
>          made, because a segment of a host domain name is now allowed
>          to begin with a digit and could legally be entirely numeric
>          (see Section 6.1.2.4).  However, a valid host name can never
>          have the dotted-decimal form #.#.#.#, since at least the
>          highest-level component label will be alphabetic.